File: typescript/packages/bench/src/metrics/paymentDelayMetric.ts
/** Mean age of the unsettled invoice pile. */

import { MetricUnit } from "../contracts/metricContract";
import type { Dataset } from "../dataset";
import { AbstractMetric } from "../support/abstractMetric";

/** Mean age of the unsettled invoice pile. */
export class PaymentDelayMetric extends AbstractMetric {
    /** Registry key, unique across the lane. */
    static readonly KEY = "payment-delay";

    constructor() {
        super(PaymentDelayMetric.KEY, MetricUnit.Days);
    }

    override compute(data: Dataset): number {
        return data.invoices.filter((invoice) => !invoice.paid).length * 3.5;
    }
}

File: rust/crates/atelier-bench/src/metrics/payment_delay_metric.rs
//! Mean age of the unsettled invoice pile.

use crate::contracts::metric_contract::{MetricContract, MetricUnit};
use crate::dataset::Dataset;
use crate::support::abstract_metric::MetricBase;

/// Mean age of the unsettled invoice pile.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct PaymentDelayMetric {
    base: MetricBase,
}

impl PaymentDelayMetric {
    /// Registry key, unique across the lane.
    pub const KEY: &'static str = "payment-delay";

    pub const fn new() -> Self {
        Self {
            base: MetricBase::new(Self::KEY, MetricUnit::Days),
        }
    }

    pub const fn base(&self) -> &MetricBase {
        &self.base
    }
}

impl Default for PaymentDelayMetric {
    fn default() -> Self {
        Self::new()
    }
}

impl MetricContract for PaymentDelayMetric {
    fn key(&self) -> &'static str {
        self.base.key
    }

    fn unit(&self) -> MetricUnit {
        self.base.unit
    }

    fn compute(&self, data: &Dataset) -> f64 {
        data.invoices.iter().filter(|invoice| !invoice.paid).count() as f64 * 3.5
    }
}
